### PR TITLE
JBS-240: Save actual snapshot size while uploading to DSS

### PR DESCRIPTION
**Pull request by shishirng**
Also, verify if downloaded snapshot size is < than expected
and error out.

If no size is saved, then we do not verify while downloading

Signed-off-by: shishir gowda <shishir.gowda@ril.com>

### CODE REVIEW

**Comment by shishirng** [comment]
Test cases executed:
1. Restore exisiting snapshots(no entry of actual_size in cinderdb)
2. create new snapshot and verify acutal_size in cinderdb
3. restore newly created snapshot

**Comment by aggarwalchirag** [comment]
Good to be merged.
Please rebase as it has merge conflicts.

File: cinder/backup/drivers/sbs.py [hunk 1]
```diff
@@ -406,7 +406,7 @@ def _get_bucket(self, conn, bucket_name):
         return backup_bucket
 
     #currently broken, not used
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Why is this comment there ? Is it relevant ? If not lets remove it.

File: cinder/backup/drivers/sbs.py [hunk 2]
```diff
@@ -419,8 +419,15 @@ def _multi_part_upload(self, bucket, key, loc):
                 mp.upload_part_from_file(fp,part_num=i+1)
 
         mp.complete_upload()
+        msg = (_('Snapshot %(snap)s Uploaded size:%(size)d' %
+                { 'snap': key, 'size': size}))
+        LOG.info(msg)
 
-    def _upload_to_DSS(self, snap_name, volume_name, ceph_args, from_snap=None):
+        self.db.backup_update(self.context, backup_id,
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Looks like this record is created before this. Is there an opportunity for us to add the chunksize during the record creation. Its okay if its not possible ?

File: cinder/backup/drivers/sbs.py [hunk 3]
```diff
@@ -449,7 +456,7 @@ def _upload_to_DSS(self, snap_name, volume_name, ceph_args, from_snap=None):
             raise exception.BackupOperationError(msg)
             return
         try:
-            self._multi_part_upload(bucket, key, loc)
+            self._multi_part_upload(bucket, key, loc, backup_id)
             #key.set_contents_from_filename(loc)
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Remove this line. If  its not required.

File: cinder/backup/drivers/sbs.py [hunk 4]
```diff
@@ -618,9 +625,11 @@ def _backup_rbd(self, backup, volume_file, volume):
                      'base': base_name})
 
         if upload_base == True:
-                self._upload_to_DSS(base_name, volume_name, ceph_args)
+                self._upload_to_DSS(base_name, volume_name, ceph_args,
+                                    base_id)
         # export diff now
-        self._upload_to_DSS(new_snap, volume_name, ceph_args, from_snap)
+        self._upload_to_DSS(new_snap, volume_name, ceph_args, backup_id,
+                            from_snap)
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Quick question on snapshot size ?
From what I see the first back does not contain the base snapshot size. 
From a metering, perspective are we including the base snapshot with the usage ?
Are we planning to handle it in the metering layer ?

File: cinder/backup/drivers/sbs.py [hunk 5]
```diff
@@ -639,14 +648,8 @@ def _backup_rbd(self, backup, volume_file, volume):
...
         self.db.backup_update(self.context, backup_id,
-			      {'created_at': now})
-        self.db.backup_update(self.context, backup_id,
-			      {'time_stamp': time_stamp})
-
-        self.db.backup_update(self.context, backup_id,
-                              {'parent_id': par_id})
-        self.db.backup_update(self.context, backup_id,
-                              {'container': self._container})
+			      {'created_at': now, 'time_stamp': time_stamp,
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Does the record already exist in the database ? If yes, 
"created_at" MUST be added at the time when the record is created. When the snapshot is complete you can add "completed_at" (very similar to the concept of "available_at" in case of volumes.

If this is the first time that record is created (which i believe its not) please change the db method name to self.db.backup_create_or_update

File: cinder/backup/drivers/sbs.py [hunk 6]
```diff
@@ -718,6 +720,25 @@ def _download_from_DSS(self, snap_name, volume_name, ceph_args, container):
...
+        actual_size = backup['actual_size']
+
+        if (actual_size != None):
+            downloaded_size = os.stat(loc).st_size
+            if (actual_size != downloaded_size):
+                errmsg = (_('Size mismatch in snapshots. Expected:%(actual)d,'
+                            ' downloaded:%(downloaded)d' %
+                           {'actual': actual_size, 'downloaded': downloaded_size}))
+                LOG.error(errmsg)
+                raise exception.InvalidBackup(reason=errmsg)
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Can you capture a metric (as discussed in scrum) when hit this case ?
Something like,
snapshot_size_mismatch = 0 for no mismatch and  snapshot_size_mismatch = 1.
Its a good idea to do it for all requests, so that you can find out what % of snapshot calls are failing.

File: cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+from oslo_log import log as logging
+from sqlalchemy import Column, MetaData, Table, Float
+
+from cinder.i18n import _LE
+
+LOG = logging.getLogger(__name__)
+
+
+def upgrade(migrate_engine):
```
Comment (souvik-ray) [inline cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py]: can u add some python docs for this methof ?

File: cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    try:
+        backups.create_column(actual_size)
+        backups.update().values(actual_size=None).execute()
+    except Exception:
+        LOG.error(_LE("Adding actual_size column to backups table failed."))
+        raise
+
+
+def downgrade(migrate_engine):
```
Comment (souvik-ray) [inline cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py]: ditto

File: cinder/backup/drivers/sbs.py [hunk 9]
```diff
@@ -693,7 +695,7 @@ def _list_incr_backups(self, backup):
         backup_tree.reverse()
         return backup_tree
 
-    def _download_from_DSS(self, snap_name, volume_name, ceph_args, container):
+    def _download_from_DSS(self, snap_name, volume_name, ceph_args, container, backup):
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: where is this method called from ?
Should not that change as well

File: cinder/backup/drivers/sbs.py [hunk 10]
```diff
@@ -664,9 +667,8 @@ def backup(self, backup, volume_file, backup_metadata=False):
         LOG.debug("Starting backup of volume='%s'.", volume_id)
 
         self.db.backup_update(self.context, backup_id,
-                              {'container': self._container})
-        self.db.backup_update(self.context, backup_id,
-                              {'version':self.DRIVER_VERSION})
+                              {'container': self._container,
+                               'version':self.DRIVER_VERSION})
```
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: just curious: what is driver version ?

File: cinder/backup/drivers/sbs.py (hunk 9 above)
Comment (aggarwalchirag) [inline cinder/backup/drivers/sbs.py]: Valid question

File: cinder/backup/drivers/sbs.py (hunk 1 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: ack,,will remove

File: cinder/backup/drivers/sbs.py (hunk 2 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: no, this size is the actual snapshot size, and is available only when we fetch it from ceph to upload to DSS

File: cinder/backup/drivers/sbs.py (hunk 4 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: base and all snapshots now contain (with this fix) 2 sizes: 1 volume size and actual snap shot size.
For metering, we can use the actual_size field

File: cinder/backup/drivers/sbs.py (hunk 10 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: we have introduced version'ing of drivers, so that we can be backward compatible. We save the driver version used to create/upload so that in the future when we evolve, older snapshots can still co-exist.

current version is '1.0'

File: cinder/backup/drivers/sbs.py (hunk 6 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: ACK.
will do that in a separate patch

File: cinder/backup/drivers/sbs.py (hunk 9 above)
Comment (shishirng) [inline cinder/backup/drivers/sbs.py]: good catch, don't know why my git commit dint have it as I had in my test env.

Will update it.

File: cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py (hunk 7 above)
Comment (shishirng) [inline cinder/db/sqlalchemy/migrate_repo/versions/046_add_actual_size_to_backups.py]: these are basically cinder-manage related scripts which add/remove columns/tables for cinder db during installation (upgrade/downgrade)

File: cinder/backup/drivers/sbs.py (hunk 5 above)
Comment (souvik-ray) [inline cinder/backup/drivers/sbs.py]: Not sure if this is acted. Creating a jira ticket
https://jira.ril.com/browse/JBS-255

As of now I am merging the code as it will have a discrepancy in the created_at field
